indexer/src/metrics.rs

```diff
@@ -0,0 +1,191 @@
+        .namespace(INDEXER_NAMESPACE)
+        .subsystem(CANDIDATES_SUBSYSTEM);
+    let num_successful = Counter::with_opts(opts)?;
+
+    registry.register(Box::new(num_successful.clone()))?;
+
+    let opts = Opts::new("num_of_failed_candidates", "Number of rejected candidates")
+        .namespace(INDEXER_NAMESPACE)
+        .subsystem(CANDIDATES_SUBSYSTEM);
+    let num_failed = Counter::with_opts(opts)?;
+
+    registry.register(Box::new(num_failed.clone()))?;
+
+    Ok(CandidatesListener {
+        num_successful,
+        num_failed,
+    })
+}
+
+pub(crate) fn make_block_listener_metrics(registry: Registry) -> Result<BlockEventListener> {
+    let opts = Opts::new("num_of_candidates", "Number of candidates indexed")
+        .namespace(INDEXER_NAMESPACE)
+        .subsystem(LISTENER_SUBSYSTEM);
+    let num_candidates = Counter::with_opts(opts)?;
+    registry.register(Box::new(num_candidates.clone()))?;
+
+    Ok(BlockEventListener { num_candidates })
+}
+
+pub(crate) fn make_publisher_metrics(registry: Registry) -> Result<PublisherListener> {
+    let opts = Opts::new("num_of_published_blocks", "Number of published blocks to MQ")
+        .namespace(INDEXER_NAMESPACE)
+        .subsystem(PUBLISHER_SUBSYSTEM);
+    let num_published_blocks = Counter::with_opts(opts)?;
+
+    registry.register(Box::new(num_published_blocks.clone()))?;
+
+    let opts = Opts::new("num_failed_published", "Number of failed publishes to MQ")
+        .namespace(INDEXER_NAMESPACE)
+        .subsystem(PUBLISHER_SUBSYSTEM);
+    let num_failed_publishes = Counter::with_opts(opts)?;
+
+    registry.register(Box::new(num_failed_publishes.clone()))?;
+
+    let publish_duration_opts = HistogramOpts::new("publish_duration_seconds", "Time spent in publishing messages")
+        .namespace(INDEXER_NAMESPACE)
+        .subsystem(PUBLISHER_SUBSYSTEM)
+        .buckets(vec![5., 10., 25., 50., 100., 250., 500., 1000., 2500., 5000., 10000.]); // ms
+    let publish_duration_histogram = Histogram::with_opts(publish_duration_opts)?;
+
+    registry.register(Box::new(publish_duration_histogram.clone()))?;
+
+    Ok(PublisherListener {
+        num_published_blocks,
+        num_failed_publishes,
+        publish_duration_histogram,
+    })
+}
+
+async fn metrics(registry: web::Data<Registry>) -> HttpResponse {
+    let metric_families = registry.gather();
+    let mut buffer = Vec::new();
+    let encoder = TextEncoder::new();
+
+    match encoder.encode(&metric_families, &mut buffer) {
+        Ok(_) => HttpResponse::Ok().content_type(encoder.format_type()).body(buffer),
+        Err(e) => HttpResponse::InternalServerError().body(e.to_string()),
+    }
+}
+
+async fn metrics_runner(metrics_addr: SocketAddr, registry: Registry) {
+    const RECONNECTION_RETRIES: u32 = 3;
+    const RECONNECTION_INTERVAL: Duration = Duration::from_secs(1);
+
+    let create_server = |registry: Registry| {
```

Comment:
There is the Rust way to solve that gracefully. Try to create an enum `MetricsRunner` with intermediate states (or wrap that in the struct, if you want to keep some data between states). And implement that like:

```rust
enum MetricsRunner {
    BindingServer,
    ServerActive,
    WaitingForReconnection,
}
```

or even just a struct (no exact methods I used, just a template):

```rust
struct MetricsRunner;

impl MetricsServer {
    pub async fn entrypoint(mut self) {
        for _ in 0..ATTEMPTS {
             if let Err(err) = self.step().await {
                 error!("Failed {err}, reconnect in {INTERVAL}");
             }
             sleep(INTERVAL).await?;
        }
    }

    async fn step(&mut self) -> Result<(), Error> {
        let server = bind_the_server().await?;
        server.run().await?;
        Ok(())
    }
}
```